### PR TITLE
Improve `fread` for very small or very large fp numbers.

### LINKED ISSUE #3492: Two test failures on armv7hl
Two tests fail on [armv7hl](https://koji.fedoraproject.org/koji/taskinfo?taskID=34042977) with 1.12.2, when they didn't on 1.12.0, tests 1018 and 1737.5:

```
Running the tests in 'tests/main.R' failed.
Last 13 lines of output:
   3: 1848  6.522     1
   4: 1835  6.181     1
   5: 1912  5.300     2
   6: 1223  4.036   728
   7:  861  3.794     1
   8: 1639  3.770   143
   9: 1253  3.723   485
  10: 1779  3.067    13
  
  Error in test.data.table() : 
    2 errors out of 6578 in 00:02:16 elapsed (00:02:02 cpu) on Mon Apr  8 03:08:59 2019. [endian==little, sizeof(long double)==8, sizeof(pointer)==4, TZ=NA, locale='C', l10n_info()='MBCS=FALSE; UTF-8=FALSE; Latin-1=FALSE', getDTthreads()='omp_get_num_procs()==4; R_DATATABLE_NUM_PROCS_PERCENT=="" (default 50); R_DATATABLE_NUM_THREADS==""; omp_get_thread_limit()==2147483647; omp_get_max_threads()==4; OMP_THREAD_LIMIT==""; OMP_NUM_THREADS==""; data.table is using 2 threads. This is set on startup, and by setDTthreads(). See ?setDTthreads.; RestoreAfterFork==true']. Search tests/tests.Rraw for test numbers: 1018, 1737.5.
  In addition: Warning message:
  In parse(n = -1, file = file, srcfile = NULL, keep.source = FALSE) :
    invalid input found on input connection '/builddir/build/BUILD/data.table/data.table.Rcheck/data.table/tests/tests.Rraw'
```

# Minimal reproducible example
```
$ _R_CHECK_FORCE_SUGGESTS_=0 R CMD check data.table
```
(xts and nanotime are not packaged yet, so I need to skip the Suggests check.)

# Output of sessionInfo()
I can't get this exactly because it's on a builder. The closest I can do is the output from a local build (though platform will be different of course):
```
R version 3.5.3 (2019-03-11)
Platform: x86_64-redhat-linux-gnu (64-bit)
Running under: Fedora 31 (Rawhide)

Matrix products: default
BLAS/LAPACK: /usr/lib64/R/lib/libRblas.so

locale:
 [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       
 [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   
 [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          
[10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

loaded via a namespace (and not attached):
[1] compiler_3.5.3
```

### LINKED ISSUE #4097: Compiler warning on platform: armv8l-unknown-linux-gnueabi (32-bit)
`
gcc I"/data/data/com.termux/files/usr/lib/R/include" -DNDEBUG   -I/data/data/com.termux/files/usr/include -fopenmp -fpic  -march=armv7-a -mfpu=neon -mfloat-abi=softfp -mthumb  -c freadR.c -o freadR.o
In file included from fread.c:28:
./freadLookups.h:31:1: warning: magnitude
      of floating-point constant too small
      for type 'long double'; minimum is
      4.9406564584124654E-324
      [-Wliteral-range]
1.0E-350L,
^
./freadLookups.h:32:1: warning: magnitude
      of floating-point constant too small
      for type 'long double'; minimum is
      4.9406564584124654E-324
      [-Wliteral-range]
1.0E-349L,
^
./freadLookups.h:33:1: warning: magnitude
      of floating-point constant too small
      for type 'long double'; minimum is
      4.9406564584124654E-324
      [-Wliteral-range]
1.0E-348L,
^
./freadLookups.h:34:1: warning: magnitude
      of floating-point constant too small
      for type 'long double'; minimum is
      4.9406564584124654E-324
      [-Wliteral-range]
1.0E-347L,
^
./freadLookups.h:35:1: warning: magnitude
      of floating-point constant too small
      for type 'long double'; minimum is
      4.9406564584124654E-324
      [-Wliteral-range]
1.0E-346L,
^
./freadLookups.h:36:1: warning: magnitude
      of floating-point constant too small
      for type 'long double'; minimum is
      4.9406564584124654E-324
      [-Wliteral-range]
1.0E-345L,
^
./freadLookups.h:37:1: warning: magnitude
      of floating-point constant too small
      for type 'long double'; minimum is
      4.9406564584124654E-324
      [-Wliteral-range]
1.0E-344L,
^
./freadLookups.h:38:1: warning: magnitude
      of floating-point constant too small
      for type 'long double'; minimum is
      4.9406564584124654E-324
      [-Wliteral-range]
1.0E-343L,
^
./freadLookups.h:39:1: warning: magnitude
      of floating-point constant too small
      for type 'long double'; minimum is
      4.9406564584124654E-324
      [-Wliteral-range]
1.0E-342L,
^
./freadLookups.h:40:1: warning: magnitude
      of floating-point constant too small
      for type 'long double'; minimum is
      4.9406564584124654E-324
      [-Wliteral-range]
1.0E-341L,
^
./freadLookups.h:41:1: warning: magnitude
      of floating-point constant too small
      for type 'long double'; minimum is
      4.9406564584124654E-324
      [-Wliteral-range]
1.0E-340L,
^
./freadLookups.h:42:1: warning: magnitude
      of floating-point constant too small
      for type 'long double'; minimum is
      4.9406564584124654E-324
      [-Wliteral-range]
1.0E-339L,
^
./freadLookups.h:43:1: warning: magnitude
      of floating-point constant too small
      for type 'long double'; minimum is
      4.9406564584124654E-324
      [-Wliteral-range]
1.0E-338L,
^
./freadLookups.h:44:1: warning: magnitude
      of floating-point constant too small
      for type 'long double'; minimum is
      4.9406564584124654E-324
      [-Wliteral-range]
1.0E-337L,
^
./freadLookups.h:45:1: warning: magnitude
      of floating-point constant too small
      for type 'long double'; minimum is
      4.9406564584124654E-324
      [-Wliteral-range]
1.0E-336L,
^
./freadLookups.h:46:1: warning: magnitude
      of floating-point constant too small
      for type 'long double'; minimum is
      4.9406564584124654E-324
      [-Wliteral-range]
1.0E-335L,
^
./freadLookups.h:47:1: warning: magnitude
      of floating-point constant too small
      for type 'long double'; minimum is
      4.9406564584124654E-324
      [-Wliteral-range]
1.0E-334L,
^
./freadLookups.h:48:1: warning: magnitude
      of floating-point constant too small
      for type 'long double'; minimum is
      4.9406564584124654E-324
      [-Wliteral-range]
1.0E-333L,
^
`

### PR DESCRIPTION
**Pull request by QuLogic**
On non-x86 architectures (armv7hl and ppc64le), test 1018 fails with a slightly differently parsed number. In base R, `R_strtod` handles small numbers by pre-dividing numerator and divisor before applying the exponent part (instead of dividing all together.) However, it does not use a lookup table.

For `fread`, trim the exponent lookup table from ±350 to ±300, and if anything is in that removed range, do two multiplications instead. This results in approximately the same effect as in base R.

Removing some of the range from the lookup table also fixes several warnings such as:

```
freadLookups.h:57:1: warning: floating constant truncated to zero [-Woverflow]
   57 | 1.0E-324L,
      | ^~~~~~~~~
freadLookups.h:690:1: warning: floating constant exceeds range of 'long double' [-Woverflow]
  690 | 1.0E309L,
      | ^~~~~~~~
```

Closes #3492 now that #4213 closed the other issue in it.
The first part of #4032 is the same issue (test 1018) fixed here. A build on those architectures can be found [here](https://koji.fedoraproject.org/koji/taskinfo?taskID=40308630).
Closes #4097 too.

### CODE REVIEW

**Comment by jangorecki** [comment]
no need to update manual? I see one example there related to precision

**Comment by QuLogic** [comment]
Which do you mean? Note this doesn't really change actual precision, just intermediary precision, so that final results are consistent.

**Comment by jangorecki** [comment]
The section in examples I meant
```
# Numerical precision :
```

**Comment by QuLogic** [comment]
Ah, I don't believe that section requires any changes. The only effect here is if the `fread` results were compared with expected values written in R code directly (and only on specific architectures), but there are no comparisons.

File: src/fread.c [hunk 1]
```diff
@@ -767,11 +767,23 @@ static void parse_double_regular(FieldParseContext *ctx)
...
 
-  double r = (double)((long double)acc * pow10lookup[e]);
-  *target = neg? -r : r;
+  long double r = (long double)acc;
+  if (e < -300 || e > 300) {
+    // Handle extra precision by pre-multiplying the result by pow(10, extra),
+    // and then remove extra from e.
+    // This avoids having to store very small or very large constants that may
+    // fail to be encoded by the compiler, even though the values can actually
+    // be stored correctly.
```
Comment (mattdowle) [inline src/fread.c]: Very clear comment and wording. Perfect.
Many thanks for investigating and fixing this!

**Pull request merged**